pageserver/src/tenant/timeline.rs

```diff
@@ -2553,6 +2711,155 @@ impl Timeline {
+        reconstruct_state: &mut ValuesReconstructState,
+        ctx: &RequestContext,
+    ) -> Result<KeySpace, GetVectoredError> {
+        let mut unmapped_keyspace = keyspace.clone();
+        let mut fringe = LayerFringe::new();
+
+        let mut completed_keyspace = KeySpace { ranges: Vec::new() };
+
+        'outer: loop {
+            if self.cancel.is_cancelled() {
+                return Err(GetVectoredError::Cancelled);
+            }
+
+            let keys_done_last_step = reconstruct_state.consume_done_keys();
+            unmapped_keyspace.remove_overlapping_with(&keys_done_last_step);
+            completed_keyspace.merge(&keys_done_last_step);
+
+            let guard = timeline.layers.read().await;
+            let layers = guard.layer_map();
+
+            let in_memory_layer = layers.find_in_memory_layer(|l| {
+                let start_lsn = l.get_lsn_range().start;
+                cont_lsn > start_lsn
+            });
+
+            match in_memory_layer {
+                Some(l) => {
+                    fringe.update(
+                        ReadableLayerDesc::InMemory {
+                            handle: l,
+                            lsn_ceil: cont_lsn,
+                        },
+                        unmapped_keyspace.clone(),
+                    );
+                }
+                None => {
+                    for range in unmapped_keyspace.ranges.iter() {
+                        let results = match layers.range_search(range.clone(), cont_lsn) {
+                            Some(res) => res,
+                            None => {
+                                break 'outer;
+                            }
+                        };
+
+                        results
+                            .found
+                            .into_iter()
+                            .map(|(res, accum)| {
+                                (
+                                    ReadableLayerDesc::Persistent {
+                                        desc: (*res.layer).clone(),
+                                        lsn_floor: res.lsn_floor,
+                                        lsn_ceil: cont_lsn,
+                                    },
+                                    accum.to_keyspace(),
+                                )
+                            })
+                            .for_each(|(layer, keyspace)| fringe.update(layer, keyspace));
+                    }
+                }
+            }
+
+            if let Some((layer_to_read, keyspace_to_read)) = fringe.next_layer() {
+                layer_to_read
+                    .get_values_reconstruct_data(
+                        &guard,
+                        keyspace_to_read.clone(),
+                        reconstruct_state,
+                        ctx,
+                    )
+                    .await?;
+
+                unmapped_keyspace = keyspace_to_read;
+                cont_lsn = layer_to_read.get_lsn_floor();
```

Comment:
Isn't this too aggressive? i.e., what if there are two L1s with the same LSN range but disjoint key range?

I guess we'd still have the other other L1 in the fringe and next loop iteration would just come here again and set the cont_lsn again to the same value?

Reply:
Right. What happens is a bit (too) subtle.

When we add a layer to the fringe we also specify the lsn range which should be read when the layer gets popped (see `lsn_ceil` and `lsn_floor` above. This means that `cont_lsn` in this function will not always decrease monotonically.

Let's take your example of two layers with the same LSN range in the fringe.

fringe = [ L1: {lsn_floor: X, lsn_ceil: Y, ..}, L2: {lsn_floor: V, lsn_ceil: W}]
pop L1: any keys that mapped to L1 and are not done will continue their search from X; let's say this happens in a layer L3
fringe = [L2: {lsn_floor: V, lsn_ceil: W}, L3: {lsn_floor: T, lsn_ceil: X}]
pop L2: now `cont_lsn` will go back in time, but that's ok because we are looking at a range with a different path

Reply:
This subtlety is actually very important, so please let me know if what I said doesn't make sense.

Reply:
Could we try to get this sublety into a comment...? It will be quite the long way around to find the PR and then find this discussion.

Reply:
It sort of is already:

```
/// This is basically a depth-first search visitor implementation where a vertex
/// is the (layer, lsn range, key space) tuple. The fringe acts as the stack.
```